File: bindings/java-jni/src/main/java/org/maplibre/nativejni/internal/callback/ResourceTransformState.java
package org.maplibre.nativejni.internal.callback;

import java.nio.charset.StandardCharsets;
import java.util.Objects;
import org.bytedeco.javacpp.BytePointer;
import org.bytedeco.javacpp.Pointer;
import org.maplibre.nativejni.internal.javacpp.JavaCppSupport;
import org.maplibre.nativejni.internal.javacpp.MaplibreNativeC;
import org.maplibre.nativejni.internal.status.Status;
import org.maplibre.nativejni.resource.ResourceKind;
import org.maplibre.nativejni.resource.ResourceTransformCallback;
import org.maplibre.nativejni.resource.ResourceTransformRequest;

/** Owns runtime-scoped resource transform callback state. */
public final class ResourceTransformState implements AutoCloseable {
  private final ResourceTransformCallback callback;
  private final MaplibreNativeC.mln_resource_transform_callback nativeCallback;
  private final MaplibreNativeC.mln_resource_transform transform;
  private final Object callbackLock = new Object();
  private final ThreadLocal<Integer> callbackDepth = ThreadLocal.withInitial(() -> 0);

  private int activeCallbacks;
  private boolean closeRequested;
  private boolean closed;

  public ResourceTransformState(ResourceTransformCallback callback) {
    this.callback = Objects.requireNonNull(callback, "callback");
    this.nativeCallback =
        new MaplibreNativeC.mln_resource_transform_callback() {
          @Override
          public int call(
              Pointer userData,
              int kind,
              BytePointer url,
              MaplibreNativeC.mln_resource_transform_response response) {
            if (!ResourceTransformState.this.enterCallback()) {
              response.url(null);
              return MaplibreNativeC.MLN_STATUS_OK;
            }
            try {
              var transformed =
                  ResourceTransformState.this.callback.transform(
                      new ResourceTransformRequest(
                          ResourceKind.fromNative(kind), JavaCppSupport.cString(url)));
              response.url(null);
              if (transformed.isPresent() && !transformed.get().isEmpty()) {
                if (transformed.get().indexOf('\0') >= 0) {
                  return MaplibreNativeC.MLN_STATUS_OK;
                }
                return setResponseUrl(response, transformed.get());
              }
              return MaplibreNativeC.MLN_STATUS_OK;
            } catch (Throwable exception) {
              response.url(null);
              return MaplibreNativeC.MLN_STATUS_NATIVE_ERROR;
            } finally {
              ResourceTransformState.this.exitCallback();
            }
          }
        };
    this.transform = new MaplibreNativeC.mln_resource_transform();
    transform.size(transform.sizeof());
    transform.callback(nativeCallback);
    transform.user_data(null);
  }

  public MaplibreNativeC.mln_resource_transform transform() {
    return transform;
  }

  public boolean isClosed() {
    synchronized (callbackLock) {
      return closed;
    }
  }

  private boolean enterCallback() {
    synchronized (callbackLock) {
      if (closeRequested || closed) {
        return false;
      }
      activeCallbacks++;
      callbackDepth.set(callbackDepth.get() + 1);
      return true;
    }
  }

  private void exitCallback() {
    synchronized (callbackLock) {
      var depth = callbackDepth.get() - 1;
      if (depth == 0) {
        callbackDepth.remove();
      } else {
        callbackDepth.set(depth);
      }
      activeCallbacks--;
      if (activeCallbacks == 0) {
        callbackLock.notifyAll();
      }
    }
  }

  @Override
  public void close() {
    var interrupted = false;
    var closeNative = false;
    synchronized (callbackLock) {
      if (callbackDepth.get() > 0) {
        throw Status.callbackReentry("Resource transform");
      }
      while (closeRequested && !closed) {
        try {
          callbackLock.wait();
        } catch (InterruptedException exception) {
          interrupted = true;
        }
      }
      if (closed) {
        if (interrupted) {
          Thread.currentThread().interrupt();
        }
        return;
      }
      closeRequested = true;
      while (activeCallbacks > 0) {
        try {
          callbackLock.wait();
        } catch (InterruptedException exception) {
          interrupted = true;
        }
      }
      closeNative = true;
    }
    try {
      if (closeNative) {
        transform.close();
        nativeCallback.close();
      }
    } finally {
      if (closeNative) {
        synchronized (callbackLock) {
          closed = true;
          callbackLock.notifyAll();
        }
      }
      if (interrupted) {
        Thread.currentThread().interrupt();
      }
    }
  }

  private static int setResponseUrl(
      MaplibreNativeC.mln_resource_transform_response response, String value) {
    var bytes = value.getBytes(StandardCharsets.UTF_8);
    try (var storage = new BytePointer(bytes.length)) {
      storage.put(bytes);
      return MaplibreNativeC.mln_resource_transform_response_set_url(
          response, storage, bytes.length);
    }
  }
}

File: bindings/kotlin/src/androidMain/kotlin/org/maplibre/nativeffi/internal/callback/ResourceTransformState.kt
package org.maplibre.nativeffi.internal.callback

import java.nio.charset.StandardCharsets
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import org.bytedeco.javacpp.BytePointer
import org.bytedeco.javacpp.Pointer
import org.maplibre.nativeffi.error.MaplibreStatus
import org.maplibre.nativeffi.internal.javacpp.JavaCppSupport
import org.maplibre.nativeffi.internal.javacpp.MaplibreNativeC
import org.maplibre.nativeffi.internal.lifecycle.HandleLeakCleaner
import org.maplibre.nativeffi.resource.ResourceKind
import org.maplibre.nativeffi.resource.ResourceTransformCallback
import org.maplibre.nativeffi.resource.ResourceTransformRequest

/** Owns runtime-scoped Android JNI resource transform callback state. */
internal class ResourceTransformState(private val callback: ResourceTransformCallback) :
  AutoCloseable {
  private val token = TOKENS.getAndIncrement()
  private val gate = CallbackGate("resource transform callbacks") { closeNative() }
  private val transform = MaplibreNativeC.mln_resource_transform()

  init {
    transform.size(transform.sizeof())
    transform.callback(NATIVE_CALLBACK)
    transform.user_data(JavaCppSupport.addressPointer(token))
    transform.release_user_data(NATIVE_RELEASE)
    STATES[token] = this
  }

  fun descriptor(): MaplibreNativeC.mln_resource_transform = transform

  fun invoke(
    rawKind: Int,
    url: BytePointer?,
    response: MaplibreNativeC.mln_resource_transform_response?,
  ): Int {
    if (response == null) return MaplibreStatus.INVALID_ARGUMENT.nativeCode
    val lease = gate.enter() ?: return MaplibreStatus.INVALID_ARGUMENT.nativeCode
    return try {
      response.size(response.sizeof())
      response.url(null)
      val replacement =
        callback.transform(
          ResourceTransformRequest(ResourceKind.fromNative(rawKind), JavaCppSupport.cString(url))
        )
      if (!replacement.isNullOrEmpty()) {
        if ('\u0000' in replacement) return MaplibreStatus.INVALID_ARGUMENT.nativeCode
        return setResponseUrl(response, replacement)
      }
      MaplibreStatus.OK.nativeCode
    } catch (_: IllegalArgumentException) {
      MaplibreStatus.INVALID_ARGUMENT.nativeCode
    } catch (_: Throwable) {
      MaplibreStatus.NATIVE_ERROR.nativeCode
    } finally {
      lease.close()
    }
  }

  fun checkCanClose() = gate.checkCanClose()

  fun isClosedForTesting(): Boolean = gate.isClosedForTesting()

  override fun close() = gate.close()

  private fun closeNative() {
    STATES.remove(token)
    transform.close()
  }

  private fun setResponseUrl(
    response: MaplibreNativeC.mln_resource_transform_response,
    value: String,
  ): Int {
    val bytes = value.toByteArray(StandardCharsets.UTF_8)
    BytePointer(bytes.size.toLong()).use { storage ->
      storage.put(bytes, 0, bytes.size)
      return MaplibreNativeC.mln_resource_transform_response_set_url(
        response,
        storage,
        bytes.size.toLong(),
      )
    }
  }

  private companion object {
    private val TOKENS = AtomicLong(1)
    private val STATES = ConcurrentHashMap<Long, ResourceTransformState>()

    /**
     * One process-wide thunk. JavaCPP's FunctionPointer pool is ten slots per generated class, so
     * per-runtime thunks ran out at eleven live runtimes.
     */
    private val NATIVE_CALLBACK =
      object : MaplibreNativeC.mln_resource_transform_callback() {
        override fun call(
          userData: Pointer?,
          kind: Int,
          url: BytePointer?,
          response: MaplibreNativeC.mln_resource_transform_response?,
        ): Int =
          STATES[userData?.address() ?: 0L]?.invoke(kind, url, response)
            ?: MaplibreStatus.INVALID_ARGUMENT.nativeCode
      }

    /** One process-wide release thunk; per-state thunks would exhaust the same pool. */
    private val NATIVE_RELEASE =
      object : MaplibreNativeC.mln_runtime_callback_release() {
        override fun call(userData: Pointer?) {
          val state = STATES[userData?.address() ?: 0L] ?: return
          HandleLeakCleaner.releaseNativeCallbackRoot(state)
          state.close()
        }
      }
  }
}
